### PR TITLE
test: new test framework for AKV provider

### PR DESCRIPTION
**Pull request by aramase**
**Reason for Change**:

- Add new test framework based on ginkgo
- Uses `e2eteam/busybox:1.29` for testing in windows. This is a smaller image.
- New test framework reduces the tests run time from 45m  to ~25m for windows clusters.
- This framework will also allow us to setup soak tests.

Follow up after PR is merged:
1. Add doc for how to run E2E tests locally
2. Add script for setting up all artifacts required for tests (cc @bdlb77)
3. Remove bats tests from the repo

**Requirements**

- [x] squashed commits
- [x] included documentation
- [x] added unit tests and e2e tests (if applicable).

**Issue Fixed**:

**Please answer the following questions with yes/no**:

- Does this change contain code from or inspired by another project? If so, did you notify the maintainers and provide attribution?

**Special Notes for Reviewers**:

### CODE REVIEW

**Comment by aramase** [comment]
@bdlb77 This is the new test framework I was working on last week. PTAL when you get a chance!

**Comment by bdlb77** [comment]
@aramase Will do! For splitting https://github.com/Azure/secrets-store-csi-driver-provider-azure/pull/45 Would you say for the testing section I should wait for this to be merged before making the PR to update the tests for local / remote container development or open a PR using bats and then a follow-up which will add any possible tests needed to make it compatible in Golang test suite?

**Comment by aramase** [comment]
> @aramase Will do! For splitting #45 Would you say for the testing section I should wait for this to be merged before making the PR to update the tests for local / remote container development or open a PR using bats and then a follow-up which will add any possible tests needed to make it compatible in Golang test suite?

@bdlb77 I think we can start with PR using bats and then follow up with a new PR for the golang test suite. WDYT?

**Comment by bdlb77** [comment]
@ara,

> @bdlb77 I think we can start with PR using bats and then follow up with a new PR for the golang test suite. WDYT?

@aramase  sounds good!

File: test/e2e/framework/certificates/certificates.go
Comment (bdlb77) [inline test/e2e/framework/certificates/certificates.go:77]: just curious with how Golang operates here... you're returning only the `return` keyword without the `err` yet for expected return types in the function declaration, you specify that there will be 2 objects that are returned `cert.Certificate, err`... does `return` interpret `err` to be the `err` output of this specific case? And where is it defined that a `cert` is returned as well?

File: test/e2e/framework/convenience.go
Comment (bdlb77) [inline test/e2e/framework/convenience.go:1]: what do you mean by `convenience.go` here with the file name?

**Review by bdlb77 — COMMENTED** [review]
looks really good and awesome to see the test suite in Go! Just have a couple of questions mainly for clarity :)

File: test/e2e/framework/certificates/certificates.go
Comment (aramase) [reply test/e2e/framework/certificates/certificates.go:77]: @bdlb77 So this is using the named return values - 

```
func X509KeyPair(certPEMBlock, keyPEMBlock, pw []byte) (cert tls.Certificate, err error) {
```

These are variables defined at the top of the function. The return statement without argument will return the named returned values.

File: test/e2e/framework/convenience.go
Comment (aramase) [reply test/e2e/framework/convenience.go:1]: I think it's mostly just helpers but I'll probably rename it to `scheme.go` as that makes the most sense.

File: test/e2e/framework/certificates/certificates.go
Comment (bdlb77) [reply test/e2e/framework/certificates/certificates.go:77]: ah neat thank you!

**Comment by aristosvo** [comment]
Just curious: these e2e tests still require a Azure KeyVault filled with secrets, keys and certs, right?

**Comment by aramase** [comment]
> Just curious: these e2e tests still require a Azure KeyVault filled with secrets, keys and certs, right?

That's right. @bdlb77 has a script in a branch that can be run to initially setup all the required artifacts for testing. Once we get this framework change updated and merged, we can add the scripts as part of the repo.

File: test/e2e/certificates_test.go
Comment (bdlb77) [inline test/e2e/certificates_test.go:67]: wow this is much cleaner with Go vs bash

File: test/e2e/framework/config.go
Comment (bdlb77) [inline test/e2e/framework/config.go:23]: I just want to make sure this wasn't an accident that you left the `KeyValue` default value still populated

**Review by bdlb77 — COMMENTED** [review]
This definitely seems much more robust that bats testing, and more thorough. Really good job

File: test/e2e/certificates_test.go
Comment (aramase) [reply test/e2e/certificates_test.go:64]: yeah 😄

File: test/e2e/framework/config.go
Comment (aramase) [reply test/e2e/framework/config.go:23]: That's just to default to a value similar to how we do it today with the bats file.

File: test/e2e/framework/helm/helm.go
Comment (aramase) [inline test/e2e/framework/helm/helm.go:118]: Current clusters are only kubenet based. I'll update the cluster config in a separate PR to use Azure CNI and then remove this configuration from testing.

File: coverage.txt
Comment (ritazh) [inline coverage.txt:1]: Did you mean to checkin this file?

File: test/e2e/auto_rotation_test.go
Comment (ritazh) [inline test/e2e/auto_rotation_test.go:232]: Where do you grant permissions to the kv resources for this identity?

File: test/e2e/framework/secretproviderclass/secretproviderclass.go
Comment (ritazh) [inline test/e2e/framework/secretproviderclass/secretproviderclass.go:23]: Might be good to create a json template for the spec to catch regressions for all the expected values in parameters. WDYT?

File: test/e2e/framework/pod/pod.go
Comment (ritazh) [inline test/e2e/framework/pod/pod.go:141]: nit: comment should be updated :)

File: test/e2e/auto_rotation_test.go
Comment (aramase) [reply test/e2e/auto_rotation_test.go:232]: Permissions and role assignments aren't done as part of the test suite. They should be done out of band as part of e2e setup.

File: test/e2e/framework/secretproviderclass/secretproviderclass.go
Comment (aramase) [reply test/e2e/framework/secretproviderclass/secretproviderclass.go:23]: Based on offline discussion:
1. We'll keep the [examples](https://github.com/Azure/secrets-store-csi-driver-provider-azure/tree/master/examples) dir which will contain sample manifests for each supported configuration.
2. For schema validation, we'll need to ensure with thorough unit testing that the AKV provider handles all expected values.

File: .pipelines/e2e-tests-template.yaml
Comment (aramase) [reply .pipelines/e2e-tests-template.yaml:73]: We don't need it for the new test framework. I'm planning to remove the old bats tests and configuration parameters in the follow up PR once we merge this and get a signal. WDYT?

File: test/e2e/Makefile
Comment (aramase) [reply test/e2e/Makefile:10]: The x509: certificate relies on legacy Common Name field. If we  use SANs in the certificate, then we don't need this. With the current artifacts we have in keyvault for testing we enable Common Name matching with `GODEBUG=x509ignoreCN=0` to resolve the error.

File: test/e2e/auto_rotation_test.go
Comment (ritazh) [reply test/e2e/auto_rotation_test.go:232]: In a followup PR, can you pls document prerequisite steps (or a script) for how someone can setup the e2e?

**Review by ritazh — APPROVED** [review]

File: test/e2e/auto_rotation_test.go
Comment (aramase) [reply test/e2e/auto_rotation_test.go:232]: sounds good! this is the issue https://github.com/Azure/secrets-store-csi-driver-provider-azure/issues/151 to track the script and docs.

**Pull request merged**
